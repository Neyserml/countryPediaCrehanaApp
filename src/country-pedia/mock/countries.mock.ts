import { Country } from "../domain/country";
import { GeographicSubRegion } from "../domain/country-pedia-response";

export const COUNTRIES_STUB: Country[] = [
  {
    name: "Afghanistan",
    area: 652230,
    capital: "Kabul",
    population: 27657145,
    alpha2Code: "AF",
    nativeName: "افغانستان",
    flag: {
      emoji: "🇦🇫",
      emojiUnicode: "U+1F1E6 U+1F1EB",
      svgFile: "https://restcountries.eu/data/afg.svg",
    },
    location: {
      latitude: 33,
      longitude: 65,
    },
    currencies: [
      {
        name: "Afghan afghani",
        symbol: "؋",
      },
    ],
    officialLanguages: [
      {
        name: "Turkmen",
        nativeName: "Türkmen",
      },
      {
        name: "Uzbek",
        nativeName: "Oʻzbek",
      },
      {
        name: "Pashto",
        nativeName: "پښتو",
      },
    ],
    callingCodes: [
      {
        name: "93",
      },
    ],
    subregion: {
      name: "Southern Asia",
      region: {
        name: "Asia",
      },
    } as GeographicSubRegion,
  },
  {
    name: "Albania",
    area: 28748,
    capital: "Tirana",
    population: 2886026,
    alpha2Code: "AL",
    nativeName: "Shqipëria",
    flag: {
      emoji: "🇦🇱",
      emojiUnicode: "U+1F1E6 U+1F1F1",
      svgFile: "https://restcountries.eu/data/alb.svg",
    },
    location: {
      latitude: 41,
      longitude: 20,
    },
    currencies: [
      {
        name: "Albanian lek",
        symbol: "L",
      },
    ],
    officialLanguages: [
      {
        name: "Albanian",
        nativeName: "Shqip",
      },
    ],
    callingCodes: [
      {
        name: "355",
      },
    ],
    subregion: {
      name: "Southern Europe",
      region: {
        name: "Europe",
      },
    } as GeographicSubRegion,
  },
  {
    name: "Angola",
    area: 1246700,
    capital: "Luanda",
    population: 25868000,
    alpha2Code: "AO",
    nativeName: "Angola",
    flag: {
      emoji: "🇦🇴",
      emojiUnicode: "U+1F1E6 U+1F1F4",
      svgFile: "https://restcountries.eu/data/ago.svg",
    },
    location: {
      latitude: -12.5,
      longitude: 18.5,
    },
    currencies: [
      {
        name: "Angolan kwanza",
        symbol: "Kz",
      },
    ],
    officialLanguages: [
      {
        name: "Portuguese",
        nativeName: "Português",
      },
    ],
    callingCodes: [
      {
        name: "244",
      },
    ],
    subregion: {
      name: "Middle Africa",
      region: {
        name: "Africa",
      },
    } as GeographicSubRegion,
  },
];
