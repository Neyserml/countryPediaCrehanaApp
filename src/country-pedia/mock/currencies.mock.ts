import { Currency } from "../domain/country-pedia-response";

export const CURRENCIES_STUB: Currency[] = [
  {
    name: "Afghan afghani",
    symbol: "؋",
    code: "AFN",
  },
  {
    name: "Euro",
    symbol: "€",
    code: "EUR",
  },
  {
    name: "Albanian lek",
    symbol: "L",
    code: "ALL",
  },
  {
    name: "Algerian dinar",
    symbol: "د.ج",
    code: "DZD",
  },
  {
    name: "United State Dollar",
    symbol: "$",
    code: "USD",
  },
  {
    name: "Angolan kwanza",
    symbol: "Kz",
    code: "AOA",
  },
  {
    name: "East Caribbean dollar",
    symbol: "$",
    code: "XCD",
  },
  {
    name: "Australian dollar",
    symbol: "$",
    code: "AUD",
  },
  {
    name: "British pound",
    symbol: "£",
    code: "GBP",
  },
];
