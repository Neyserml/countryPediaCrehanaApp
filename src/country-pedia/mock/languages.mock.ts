import { Language } from "../domain/country-pedia-response";

export const LANGUAGES_STUB: Language[] = [
  {
    name: "Pashto",
    nativeName: "پښتو",
  },
  {
    name: "Uzbek",
    nativeName: "Oʻzbek",
  },
  {
    name: "Turkmen",
    nativeName: "Türkmen",
  },
  {
    name: "Swedish",
    nativeName: "svenska",
  },
  {
    name: "Albanian",
    nativeName: "Shqip",
  },
  {
    name: "Arabic",
    nativeName: "العربية",
  },
  {
    name: "English",
    nativeName: "English",
  },
  {
    name: "Samoan",
    nativeName: "gagana fa'a Samoa",
  },
  {
    name: "Catalan",
    nativeName: "català",
  },
  {
    name: "Portuguese",
    nativeName: "Português",
  },
];
