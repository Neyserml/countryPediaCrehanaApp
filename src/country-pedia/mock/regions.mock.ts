import { Region } from "../domain/country-pedia-response";

export const REGIONS_STUB: Region[] = [
  {
    name: "Asia",
    subregions: [
      {
        name: "Central Asia",
      },
      {
        name: "Eastern Asia",
      },
      {
        name: "South-Eastern Asia",
      },
      {
        name: "Western Asia",
      },
      {
        name: "Southern Asia",
      },
    ],
  },
  {
    name: "Europe",
    subregions: [
      {
        name: "Eastern Europe",
      },
      {
        name: "Western Europe",
      },
      {
        name: "Southern Europe",
      },
      {
        name: "Northern Europe",
      },
    ],
  },
  {
    name: "Africa",
    subregions: [
      {
        name: "Eastern Africa",
      },
      {
        name: "Southern Africa",
      },
      {
        name: "Western Africa",
      },
      {
        name: "Middle Africa",
      },
      {
        name: "Northern Africa",
      },
    ],
  },
  {
    name: "Oceania",
    subregions: [
      {
        name: "Micronesia",
      },
      {
        name: "Melanesia",
      },
      {
        name: "Australia and New Zealand",
      },
      {
        name: "Polynesia",
      },
    ],
  },
  {
    name: "Americas",
    subregions: [
      {
        name: "Northern America",
      },
      {
        name: "Central America",
      },
      {
        name: "South America",
      },
      {
        name: "Caribbean",
      },
    ],
  },
  {
    name: "Polar",
    subregions: [
      {
        name: "",
      },
    ],
  },
];
