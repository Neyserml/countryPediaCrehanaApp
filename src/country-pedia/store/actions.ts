interface DispatchObject<T, P = any> {
  type: T;
  payload?: P;
}

export enum CountryPediaAction {
  ClearState = "CLEAR_STATE",
  SetLanguages = "SET_LANGUAGES",
  SetCurrencies = "SET_CURRENCIES",
  SetRegions = "SET_REGIONS",
  SetCountries = "SET_COUNTRIES",
  SetCountryId = "SET_COUNTRY_ID",
}

export type CountryPediaDispatchObject<T> = DispatchObject<
  CountryPediaAction,
  T
>;
