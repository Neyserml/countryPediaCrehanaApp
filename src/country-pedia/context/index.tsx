import { castDraft } from "immer";
import React, { createContext, PropsWithChildren, useReducer } from "react";
import { DispatchState } from "../../shared/hooks/useDispatchState";
import { CountryPediaDispatchObject } from "../store/actions";
import {
  countryPediaContextReducer,
  CountryPediaState,
  initialState,
} from "../store/reducer";

export type CountryPediaDispatch<T = any> = React.Dispatch<
  CountryPediaDispatchObject<T>
>;

export type CountryPediaContextType = DispatchState<
  CountryPediaState,
  CountryPediaDispatch
>;

export const CountryPediaContext = createContext({} as CountryPediaContextType);

type CountryPediaProviderProps = PropsWithChildren<{
  extraState?: Partial<CountryPediaState>;
}>;

export const CountryPediaProvider: React.FC<CountryPediaProviderProps> = ({
  children,
  extraState = {},
}: CountryPediaProviderProps) => {
  const [state, dispatch] = useReducer(countryPediaContextReducer, {
    ...initialState,
    ...extraState,
  });

  const value = { dispatch, state: castDraft(state) };
  return (
    <CountryPediaContext.Provider value={value}>
      {children}
    </CountryPediaContext.Provider>
  );
};
