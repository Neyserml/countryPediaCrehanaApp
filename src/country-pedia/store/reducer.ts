import produce, { Draft } from "immer";
import { Country } from "../domain/country";
import { Currency, Language, Region } from "../domain/country-pedia-response";
import { CountryPediaAction, CountryPediaDispatchObject } from "./actions";

export interface CountryPediaState {
  title: string;
  languages: Language[] | null;
  currencies: Currency[] | null;
  regions: Region[] | null;
  countries: Country[] | null;
  countryId: string;
}

export const initialState: CountryPediaState = {
  title: "CountryPedia",
  languages: null,
  currencies: null,
  regions: null,
  countries: null,
  countryId: "",
};

export const countryPediaContextReducer = produce(
  (
    draft: Draft<CountryPediaState>,
    { type, payload }: CountryPediaDispatchObject<any>
  ) => {
    switch (type) {
      case CountryPediaAction.ClearState:
        return initialState;
      case CountryPediaAction.SetLanguages:
        draft.languages = payload;
        break;
      case CountryPediaAction.SetCurrencies:
        draft.currencies = payload;
        break;
      case CountryPediaAction.SetRegions:
        draft.regions = payload;
        break;
      case CountryPediaAction.SetCountries:
        draft.countries = payload;
        break;
      case CountryPediaAction.SetCountryId:
        draft.countryId = payload;
        break;
      default:
        throw new Error(`Unknown countryPedia action type: ${type}`);
    }
  }
);
