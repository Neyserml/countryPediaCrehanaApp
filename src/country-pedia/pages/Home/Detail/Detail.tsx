import React from "react";
import { useHistory } from "react-router-dom";
import styled from "styled-components";
import CardTemplate from "../../../../shared/components/CardTemplate/CardTemplete";
import Header from "../../../../shared/components/Header/Header";
import useDispatchState from "../../../../shared/hooks/useDispatchState";
import { CountryPediaContext } from "../../../context";

const StayledTitle = styled.h4`
  text-align: center;
`;

const Detail: React.FC = () => {
  const [state] = useDispatchState(CountryPediaContext);
  const history = useHistory();

  const countriesDetail = state.countries?.filter(
    ({ name }) => state.countryId === name
  );

  const onBack = (): void => {
    history.goBack();
  };

  return (
    <>
      <Header />

      <div className="container">
        <StayledTitle>{` Detalle sobre ${state.countryId}`}</StayledTitle>

        {countriesDetail && (
          <div className="row">
            <div className="col s12 m6 offset-m3 ">
              <CardTemplate
                onClickDetail={() => {}}
                title={countriesDetail[0].name}
                img={countriesDetail[0].flag.svgFile}
              >
                <a onClick={onBack} href="#">
                  Volver
                </a>
                <table>
                  <tbody>
                    <tr>
                      <td>
                        <b>Capital: </b>
                        {countriesDetail[0].capital}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Nombre Nativo: </b>
                        {countriesDetail[0].nativeName}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Población: </b>
                        {countriesDetail[0].population}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Área: </b>
                        {`${countriesDetail[0].area} Km²`}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Ubicación </b>
                        <br />
                        {`Latitud: ${countriesDetail[0].location.latitude}`}
                        <br />
                        {`Longitud: ${countriesDetail[0].location.longitude}`}
                      </td>
                    </tr>
                  </tbody>
                </table>
                <br />
              </CardTemplate>
            </div>
          </div>
        )}
      </div>
    </>
  );
};

export default Detail;
