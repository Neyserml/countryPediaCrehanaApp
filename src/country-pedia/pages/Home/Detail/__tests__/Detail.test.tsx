import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { CountryPediaProvider } from "../../../../context";
import { COUNTRIES_STUB } from "../../../../mock/countries.mock";
import { CountryPediaState, initialState } from "../../../../store/reducer";
import Detail from "../Detail";

const mockHistoryGoBack = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    goBack: mockHistoryGoBack,
  }),
}));

const state: CountryPediaState = {
  ...initialState,
  countries: COUNTRIES_STUB,
  countryId: "Albania",
};

const renderComponent = (extraState = state) => {
  return render(
    <CountryPediaProvider extraState={extraState}>
      <Detail />
    </CountryPediaProvider>
  );
};

describe("Detail Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match snapshot in detail page", () => {
    const component = renderComponent();
    expect(component.container).toMatchSnapshot();
  });

  it("should go back", () => {
    renderComponent();
    const link = screen.getByText("Volver");
    fireEvent.click(link);
    expect(mockHistoryGoBack).toHaveBeenCalled();
  });
});
