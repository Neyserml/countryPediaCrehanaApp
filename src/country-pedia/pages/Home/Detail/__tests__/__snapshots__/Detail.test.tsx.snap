// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Component should match snapshot in detail page 1`] = `
<div>
  <nav
    class="sc-dlfnbm etGDIC"
  >
    <nav
      class="nav-wrapper"
    >
      <a
        class="brand-logo"
        href="#"
      >
        Crehana
      </a>
    </nav>
  </nav>
  <div
    class="container"
  >
    <h4
      class="sc-hKgILt ZTQiC"
    >
       Detalle sobre Albania
    </h4>
    <div
      class="row"
    >
      <div
        class="col s12 m6 offset-m3 "
      >
        <div
          class="sc-gsTCUz fwTMXJ"
        >
          <div
            class="card"
          >
            <div
              class="card-image waves-effect waves-block waves-light"
            >
              <img
                class="sc-bdfBwQ iPjGkV"
                src="https://restcountries.eu/data/alb.svg"
              />
            </div>
            <div
              class="card-content"
            >
              <span
                class="card-title activator grey-text text-darken-4"
              >
                <b>
                  Albania
                </b>
              </span>
              <a
                href="#"
              >
                Volver
              </a>
              <table>
                <tbody>
                  <tr>
                    <td>
                      <b>
                        Capital: 
                      </b>
                      Tirana
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <b>
                        Nombre Nativo: 
                      </b>
                      Shqipëria
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <b>
                        Población: 
                      </b>
                      2886026
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <b>
                        Área: 
                      </b>
                      28748 Km²
                    </td>
                  </tr>
                  <tr>
                    <td>
                      <b>
                        Ubicación 
                      </b>
                      <br />
                      Latitud: 41
                      <br />
                      Longitud: 20
                    </td>
                  </tr>
                </tbody>
              </table>
              <br />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
